tests/Model/BSONArrayTest.php

```diff
@@ -17,6 +17,24 @@ public function testBsonSerializeReindexesKeys()
         $this->assertSame(['foo', 'bar'], $array->bsonSerialize());
     }
 
+    public function testClone()
+    {
+        $arr1 = new BSONArray(['x' => 1]);
+        $arr2 = new BSONArray(['a', 'b', 'c']);
+        $arr3 = new BSONArray(['foo' => 'bar']);
```

Comment:
`$arr1` and `$arr3` are technically documents, since you're using associative arrays. Perhaps we can use the following as our object:

```
new BSONArray([
    // array of multiple types (scalar, object, array)
    [
        'foo',
        new stdClass,
        ['bar', new stdClass],
    ],
    // BSONArray of multiple types (scalar, object, array)
    new BSONArray([
        'foo',
        new stdClass,
        ['bar', new stdClass],
    ]),
]);
```

The second element should cover nested BSONArray objects.

While `assertSameDocument()` should work just fine on a BSONArray and its clone, you can optionally wrap it in an associative array first (e.g. `['array' => $bsonArrayBeingTested]`).

After the cloning, you'd then want to assert objects are different in the following indices:

 * `$document['array'][0][1]`
 * `$document['array'][0][2][1]`
 * `$document['array'][1]`
 * `$document['array'][1][1]`
 * `$document['array'][1][2][1]`